### PR TITLE
Use @timestamp field to route documents to a backing index of a data stream

### PR DESCRIPTION
**Pull request by martijnvg**
Currently documents that target data streams are resolved to the target the write index of the data stream being targeted.
This change adjust this logic in the bulk api, to first parse the `@timestamp` field and then based on this timestamp select the right backing index. If the parsed timestamp of a document falls between a backing index's start_time and end_time then this backing index is used as write index. 

Note that this logic is only enabled for tsdb data streams. A temporal slice of backing indices never overlap within a data stream, so either 1 backing index can be selected or none.  

Relates #74660

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java [hunk 1]
```diff
@@ -718,6 +733,40 @@ public boolean isRequireAlias() {
...
+        if (dataStream.isTimeSeries(metadata::index) == false) {
+            return ia.getWriteIndex();
+        }
+
+        try (XContentParser parser = contentType.xContent().createParser(TS_EXTRACT_CONFIG, source().streamInput())) {
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+            ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.nextToken(), parser);
+            ensureExpectedToken(XContentParser.Token.VALUE_STRING, parser.nextToken(), parser);
+            String timestampAsString = parser.text();
+            // TODO: deal with nanos too here.
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/index/IndexRequest.java:756]: We add something (format string) to `MappingMetadata` that indicates how the `@timestamp` field should be parsed. We would need to fetch this from the latest backing index of a data stream.

Alternatively we can add information of how the `@timestamp` field should be parsed to `DataStream` class. Which feels like a better place, since we know this prior to selecting the right backing index based on the `@timestamp` field here.

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java [hunk 2]
```diff
@@ -718,6 +733,40 @@ public boolean isRequireAlias() {
...
+            // (the index hasn't been resolved yet, keep track of timestamp field metadata at data stream level, so we can use it here)
+            Instant timestamp = Instant.from(DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.parse(timestampAsString));
+
+            Index result = dataStream.selectWriteIndex(timestamp, metadata);
+            if (result == null) {
+                throw new IllegalArgumentException("no index available for a document with an @timestamp of [" + timestampAsString + "]");
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/action/index/IndexRequest.java:762]: It would be great if we could add a bit more useful information here. I think I would rephrase it as "the document timestamp [2022-01-07T19:04:41Z] is outside of ranges of currently writable indices: [[2022-01-06T00:00:00.000Z-2022-01-06T16:02:12.251Z], [2022-01-06T16:02:12.251Z-2022-01-06T20:02:12.251Z]]" or something like this.

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java [hunk 3]
```diff
@@ -718,6 +733,40 @@ public boolean isRequireAlias() {
         return requireAlias;
     }
 
+    @Override
+    public Index getConcreteWriteIndex(IndexAbstraction ia, Metadata metadata) {
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/action/index/IndexRequest.java:737]: It feels like a bit too much logic for a data class, that the IndexRequest is essentially is. I wonder if it makes more sense as a part of IndexAbstraction instead.

**Review by imotov — COMMENTED** [review]
Looks good to me in general from the TSDB perspective. Left a couple of suggestions.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java [hunk 4]
```diff
@@ -171,6 +175,32 @@ public Index getWriteIndex() {
         return indices.get(indices.size() - 1);
     }
 
+    public Index selectWriteIndex(Instant timestamp, Metadata metadata) {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:178]: Can you rename this to be something TSDB specific, like `selectTimeseriesWriteIndex` (since we may end up adding different selection criteria in the future) and add javadocs?

File: server/src/test/java/org/elasticsearch/action/index/IndexRequestTests.java [hunk 5]
```diff
@@ -272,4 +281,102 @@ public void testRejectsEmptyStringPipeline() {
         assertThat(validate, notNullValue());
         assertThat(validate.getMessage(), containsString("pipeline cannot be an empty string"));
     }
+
+    public void testGetConcreteWriteIndex() {
```
Comment (dakrone) [inline server/src/test/java/org/elasticsearch/action/index/IndexRequestTests.java:285]: Can you add a test that uses epoch millis for the `@timestamp` instead of a string to ensure that it picks the right backing index also?

**Review by dakrone — APPROVED** [review]
I left a few really minor comments, but otherwise this looks good to me.

> A temporal slice of backing indices never overlap within a data stream, so either 1 backing index can be selected or none.

Can you point me to where we do this validation? I wanted to get more familiar with it.

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 3 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/action/index/IndexRequest.java:725]: Implemented via: f97d88935098460c2f1a7cfafb6d4271636b60a5

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 2 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: Also implemented via: f97d88935098460c2f1a7cfafb6d4271636b60a5

File: server/src/main/java/org/elasticsearch/action/DocWriteRequest.java [hunk 6]
```diff
@@ -151,6 +154,10 @@
      */
     int route(IndexRouting indexRouting);
 
+    default Index getConcreteWriteIndex(IndexAbstraction ia, Metadata metadata) {
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/action/DocWriteRequest.java:165]: Added via: efdd833ef361d8e1ee153c785d384ec89e78e796

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java (hunk 4 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java]: Renamed and add jdocs: 36af27e2bdfaa9d58d9647265574e9754c8fd373

File: server/src/test/java/org/elasticsearch/action/index/IndexRequestTests.java (hunk 5 above)
Comment (martijnvg) [reply server/src/test/java/org/elasticsearch/action/index/IndexRequestTests.java:286]: I added a test for this and also added logic for this. The case of providing timestamp as a number representing mills since epoch failed. I fixes this via: d70aa28fc45bbf886b30584a753ccdc37f260343

**Comment by martijnvg** [comment]
> Can you point me to where we do this validation? I wanted to get more familiar with it.

The validation that validates the start and end time settings across backing indices doesn't yet exist.
I think this validation should be added to `Metadata.Builder#build()` method. This place we have know all data streams and we access to all IndexMetadata instances just before a new metadata instance comes into effect.

But maybe we can do this differently, now that we are going to make time series typed data streams. So I like to add this validation after we made data streams aware of index modes.

**Review by imotov — APPROVED** [review]
LGTM, Thanks!

**Review by dakrone — APPROVED** [review]
LGTM also!

> So I like to add this validation after we made data streams aware of index modes.

Makes sense, thanks for the heads up
